src/main/scala/org/camunda/feel/impl/parser/FeelParser.scala

```diff
@@ -16,475 +16,505 @@
+  private def simpleExpressions[_ : P]: P[ConstList] =
+    P[ConstList](simpleExpression.rep(1, sep = ",").map(s => ConstList(s.toList)))
 
   // 5
-  private lazy val simpleExpression
-    : Parser[Exp] = arithmeticExpression | simpleValue
+  private def simpleExpression[_ : P]: P[Exp] = P[Exp](arithmeticExpression | simpleValue)
 
   // 4 a) -> 21+22
-  private lazy val arithmeticExpression =
-    chainl1(arithmeticExpression2, "+" ^^^ Addition | "-" ^^^ Subtraction)
+  private def arithmeticExpression[_ : P] = P(
+    chainl1(arithmeticExpression2, P("+") ^^^ Addition | P("-") ^^^ Subtraction)
+  )
   // 4 b) -> 23+24
-  private lazy val arithmeticExpression2 =
-    chainl1(arithmeticExpression3, "*" ^^^ Multiplication | "/" ^^^ Division)
+  private def arithmeticExpression2[_ : P] = P(
+    chainl1(arithmeticExpression3, P("*") ^^^ Multiplication | P("/") ^^^ Division)
+  )
   // 4 c) -> 25
-  private lazy val arithmeticExpression3 =
-    chainl1(arithmeticExpression4, "**" ^^^ Exponentiation)
+  private def arithmeticExpression3[_ : P] = P(chainl1(arithmeticExpression4, P("**") ^^^ Exponentiation))
   // 4 d) -> 26
-  private lazy val arithmeticExpression4 = opt("-") ~ expression6 ^^ {
-    case Some(_) ~ e => ArithmeticNegation(e)
-    case None ~ e    => e
-  }
+  private def arithmeticExpression4[_ : P] =
+    P(("-".!.? ~ expression6).map {
+      case (Some(_), e) => ArithmeticNegation(e)
+      case (None, e) => e
+    })
 
   // 17
-  private lazy val unaryTests: Parser[Exp] =
-    "not" ~! "(" ~> positiveUnaryTests <~ ")" ^^ Not |
+  private def unaryTests[_ : P]: P[Exp] = P[Exp](
+    ("not" ~ "(" ~ positiveUnaryTests ~ ")").map(Not) |
       positiveUnaryTests |
-      "-" ^^^ ConstBool(true)
+      P("-") ^^^ ConstBool(true)
+  )
 
   // 16
-  private lazy val positiveUnaryTests: Parser[Exp] = rep1sep(positiveUnaryTest,
-                                                             ",") ^^ {
-    case test :: Nil => test
-    case tests       => AtLeastOne(tests)
-  }
+  private def positiveUnaryTests[_ : P]: P[Exp] =
+    P[Exp](positiveUnaryTest.rep(1, ",").map {
+      case Seq(test) => test
+      case tests => AtLeastOne(tests.toList)
+    })
 
   // 15 - in DMN 1.2 it's only 'expression' (which also covers simple positive unary test)
   //    - however, parse simple positive unary test first since this is most usual
-  private lazy val positiveUnaryTest: Parser[Exp] = (
-    "null" ^^^ InputEqualTo(ConstNull) |
-      simplePositiveUnaryTest |
-      expression ^^ (UnaryTestExpression(_))
-  )
+  private def positiveUnaryTest[_ : P]: P[Exp] =
+    P[Exp](
+      P("null") ^^^ InputEqualTo(ConstNull) |
+        // TODO: This makes it seemingly impossible for fastparse to backtrack its way out
+        //  when it should be using the expression.map part.
+        //  From the comment above it should not be necessary.
+        // simplePositiveUnaryTest |
+        expression.map(UnaryTestExpression)
+    )
```

Comment:
It seems that the backtracking fails when evaluating the `identifier` parser :thinking: 

However, we need to parse `simplePositiveUnaryTest` before `expression` to fulfil the boolean test cases in `InterpreterUnaryTest`. Otherwise, a boolean value is not parsed as an `InputEqualTo(ValBoolean())` expression. Instead, it is parsed as `UnaryTestExpression(ValBoolean())` and evaluates to `true` if the boolean value is `true`. See my comment on `InterpreterUnaryTest`.

We can make it work by modifying the `simpleValue` parser:
```
private def simpleValue[_: P] =
    P(
      simpleLiteral |
        !(functionInvocation | builtinFunctionInvocation) ~ qualifiedName.map(
          Ref(_)) |
        qmark
    )
```